ctl/string.cc

```diff
@@ -96,41 +115,54 @@ string::resize(size_t n2, char ch) noexcept
     if (ckd_add(&c2, n2, 1))
         __builtin_trap();
     reserve(c2);
-    if (n2 > n)
-        memset(p + n, ch, n2 - n);
-    p[n = n2] = 0;
+    if (n2 > size())
+        memset(data() + size(), ch, n2 - size());
+    if (isbig()) {
+        big()->p[big()->n = n2] = 0;
+    } else {
+        set_small_size(n2);
+        data()[size()] = 0;
+    }
 }
 
 void
 string::append(char ch) noexcept
 {
-    if (n + 2 > c) {
-        size_t c2 = c + 2;
-        c2 += c2 >> 1;
+    size_t n2;
+    if (ckd_add(&n2, size(), 2))
+        __builtin_trap();
+    if (n2 > capacity()) {
+        size_t c2 = capacity() + 2;
+        if (ckd_add(&c2, c2, c2 >> 1))
+            __builtin_trap();
         reserve(c2);
     }
-    p[n++] = ch;
-    p[n] = 0;
+    // XXX do we care to fence this?
```

Comment:
The fence is used because if a trap or asynchronous signal is raised, and I want to `pthread_exit()` the thread from the signal handler as a result of that, I'd like to be able to have my pthread cleanup callbacks free these strings, or alternatively, re-use the objects for the next request. So for example, don't say this:

```
if (!(this->p = realloc(this->p, c2)))
  __builtin_trap();
```

Because then the signal handler can't free the original allocation. The object needs to be in a reasonable state when trap is called, or an exception is thrown, otherwise I want to know if that's not possible. By reasonable I mean good enough to ~destroy or clear() and go at it again.

Reply:
This is pretty neat stuff you guys are doing.

I'm assuming the issue here is not to rewrite `this->p` before knowing `realloc(this->p, ...)` succeeds. Would just using `reallocf` solve the problem? I am also unsure as to whether the standard specifies that the original memory allocation (ie. first argument to realloc) is guaranteed to remain valid when realloc fails.

Reply:
Yes. It's an ironclad guarantee that when `realloc()` fails, the original memory is left unharmed. I haven't heard of `reallocf()` until now. It sounds similar to `realloc_in_place()` which is likely its modern replacement, which has good semantics. Really happy to see you here btw!

Reply:
Indeed, good to see you @ghaerr!

In any case, resolving this conversation as for this PR's purposes the answer to the question is simply "no."